### PR TITLE
Microsoft: Update Linux/x86 Kernel config for 4.19.121

### PR DESCRIPTION
**Pull request by microhobby**
For use with the latest version, we need the updated config.
This also can be used as a base to build the mainline version.

Solve issues:

microsoft/WSL#5366
microsoft/WSL#5481
#142

### CODE REVIEW

**Comment by WSLUser** [comment]
So I hate to tell you this but if you look at the drop down menu for branches, they do keep it up to date. It's `master` that isn't. Also WSL2 is Hyper-V based, so KVM config stuff is completely N/A (supporting use of KVM is completely different than optimizing for a KVM based VM, which WSL2 is not). See my PR for more on that (a non-MS reviewer kindly explained about KVM). This is making changes that don't really make much sense without some explanations. I've recently updated my config and now is not only compilable and bootable but actually improves things (and is based on the latest 5.7 series). There are more minor PRs in existence and while this one is still smaller, it's not by much. I highly suggest looking at the config options I enabled and then seeing what's missing and doing your compilation with those changes.

File: Microsoft/config-wsl [hunk 1]
```diff
@@ -2102,7 +2136,23 @@ CONFIG_CLKEVT_I8253=y
 CONFIG_I8253_LOCK=y
 CONFIG_CLKBLD_I8253=y
 # CONFIG_MAILBOX is not set
-# CONFIG_IOMMU_SUPPORT is not set
+CONFIG_IOMMU_API=y
```
Comment (WSLUser) [inline Microsoft/config-wsl:2139]: Is this even supported in WSL2? Hyper-V supports this but WSL2 is a lightweight version of that and does not have all the same functionalities (part of why it's lightweight).

File: Microsoft/config-wsl [hunk 2]
```diff
@@ -2469,11 +2519,11 @@ CONFIG_SECURITY=y
 # CONFIG_SECURITY_NETWORK is not set
 CONFIG_PAGE_TABLE_ISOLATION=y
 # CONFIG_SECURITY_PATH is not set
+# CONFIG_INTEL_TXT is not set
 CONFIG_HAVE_HARDENED_USERCOPY_ALLOCATOR=y
 # CONFIG_HARDENED_USERCOPY is not set
 # CONFIG_FORTIFY_SOURCE is not set
-CONFIG_STATIC_USERMODEHELPER=y
```
Comment (WSLUser) [inline Microsoft/config-wsl:2475]: This will break something in compilation or boot by disabling this.

File: Microsoft/config-wsl [hunk 3]
```diff
@@ -1952,6 +1974,7 @@ CONFIG_BCMA_POSSIBLE=y
 #
 # AMD Library routines
 #
+CONFIG_DXGKRNL=y
```
Comment (WSLUser) [inline Microsoft/config-wsl:1977]: Is this supposed be from the new CUDA/GPU support? That would require using the kernel in the 4.19.x branch. For future reference, it's been stated several times now in random Tweets that they will be shifting to the 5.4.x series. So I expect that support to be added to mainline kernel before it trickles down to 5.4.x.

File: Microsoft/config-wsl [hunk 4]
```diff
@@ -358,6 +359,9 @@ CONFIG_X86_SMAP=y
 CONFIG_X86_INTEL_UMIP=y
 CONFIG_X86_INTEL_MPX=y
 CONFIG_X86_INTEL_MEMORY_PROTECTION_KEYS=y
+CONFIG_X86_INTEL_TSX_MODE_OFF=y
+# CONFIG_X86_INTEL_TSX_MODE_ON is not set
+# CONFIG_X86_INTEL_TSX_MODE_AUTO is not set
```
Comment (WSLUser) [inline Microsoft/config-wsl:364]: Auto will turn on or off autamatically. This should not be  changed

File: Microsoft/config-wsl [hunk 5]
```diff
@@ -556,8 +560,23 @@ CONFIG_RESET_ATTACK_MITIGATION=y
...
+CONFIG_HAVE_KVM_IRQ_ROUTING=y
+CONFIG_HAVE_KVM_EVENTFD=y
+CONFIG_KVM_MMIO=y
+CONFIG_KVM_ASYNC_PF=y
+CONFIG_HAVE_KVM_MSI=y
+CONFIG_HAVE_KVM_CPU_RELAX_INTERCEPT=y
+CONFIG_KVM_VFIO=y
+CONFIG_KVM_GENERIC_DIRTYLOG_READ_PROTECT=y
+CONFIG_KVM_COMPAT=y
+CONFIG_HAVE_KVM_IRQ_BYPASS=y
```
Comment (WSLUser) [inline Microsoft/config-wsl:574]: All of these need validation they actually contribute something towards kvm support and NOT KVM guess VM support.

**Comment by microhobby** [comment]
> So I hate to tell you this but if you look at the drop down menu for branches, they do keep it up to date. It's `master` that isn't. Also WSL2 is Hyper-V based, so KVM config stuff is completely N/A (supporting use of KVM is completely different than optimizing for a KVM based VM, which WSL2 is not). See my PR for more on that (a non-MS reviewer kindly explained about KVM). This is making changes that don't really make much sense without some explanations. I've recently updated my config and now is not only compilable and bootable but actually improves things (and is based on the latest 5.7 series). There are more minor PRs in existence and while this one is still smaller, it's not by much. I highly suggest looking at the config options I enabled and then seeing what's missing and doing your compilation with those changes.

Nope, on master we have the configs for 4.19.81: https://github.com/microsoft/WSL2-Linux-Kernel/blob/master/Microsoft/config-wsl

Funny that you say about KVM, the KVM configs here are to use the `nestedVirtualization=true` on `.wslconfig` I think. Anyway, this PR is based on the Kernel `/proc/config.gz` that came from Windows Update and is exactly this `.config` that they used in the Kernel `4.19.121`.

I also used this as base for build 5.8.0-rc4 and everything works very well.

**Comment by microhobby** [comment]
I do not expect this to be merged, but I am pointing here in case anyone has the issues mentioned in the first comment.

**Comment by WSLUser** [comment]
Yes I did some more research into KVM optimizations that the host kernel should enable and many of what you have are in fact there. I added them on my personal config but ran into some issues. I also tried adding TLS from menuconfig as well and that seemed to break networking again. It seems networking is a very sensitive thing when making kernel changes with it.

**Comment by WSLUser** [comment]
Can you bump your fork to 5.4? I'd like to see to the updated config they put so I can update my fall-back to 5.4 (They haven't updated the msi).

**Pull request closed without merging**